build/pause/Makefile

```diff
@@ -12,42 +12,65 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-.PHONY:	build push
+.PHONY: all push push-legacy-too container clean
 
-TAG=2.0
-REGISTRY?="gcr.io/google_containers"
+REGISTRY ?= "gcr.io/google_containers"
+IMAGE = gcr.io/google_containers/pause-$(ARCH)
+TAG = 3.0
 
 # Architectures supported: amd64, arm, arm64 and ppc64le
-ARCH?=amd64
-GOLANG_VERSION=1.6
-TEMP_DIR:=$(shell mktemp -d)
+ARCH ?= amd64
 
-all: push-legacy-too
+CFLAGS = -Os -Wall -static
+KUBE_CROSS_VERSION ?= $(shell cat ../build-image/cross/VERSION)
 
-build:
-	cp pause.go Dockerfile $(TEMP_DIR)
+BIN = pause
+SRCS = pause.c
 
-	docker run -it -v $(TEMP_DIR):/build golang:$(GOLANG_VERSION) /bin/bash -c \
-		"cd /build && CGO_ENABLED=0 GOARM=6 GOARCH=$(ARCH) go build -a -installsuffix cgo -ldflags '-w' ./pause.go"
+ifeq ($(ARCH),amd64)
+	TRIPLE ?= x86_64-linux-gnu
+endif
 
-ifeq ($(ARCH),$(filter $(ARCH),amd64 arm))
-# Run goupx for amd64 and arm. The condition above is equal to 'if [[ $ARCH == "amd64" || $ARCH == "arm" ]]' in bash
-	docker run -it -v $(TEMP_DIR):/build golang:$(GOLANG_VERSION) /bin/bash -c \
-		"cd /build && apt-get update && apt-get install -y upx \
-		&& go get github.com/pwaller/goupx && goupx pause"
+ifeq ($(ARCH),arm)
+	TRIPLE ?= arm-linux-gnueabi
 endif
 
-	# And build the image
-	docker build -t $(REGISTRY)/pause-$(ARCH):$(TAG) $(TEMP_DIR)
+ifeq ($(ARCH),arm64)
+	TRIPLE ?= aarch64-linux-gnu
+endif
+
+ifeq ($(ARCH),ppc64le)
+	TRIPLE ?= powerpc64le-linux-gnu
+endif
 
+TEMP_DIR := $(shell mktemp -d)
+
+all: $(TEMP_DIR)/.push $(TEMP_DIR)/.push-legacy-too
+
+$(TEMP_DIR)/$(BIN): $(SRCS)
+	cp $(SRCS) Dockerfile $(TEMP_DIR)
+	docker run -v $(TEMP_DIR):/build gcr.io/google_containers/kube-cross:$(KUBE_CROSS_VERSION) \
+		/bin/bash -c "\
+			cd /build && \
+			$(TRIPLE)-gcc $(CFLAGS) -o $(BIN) $^ && \
+			$(TRIPLE)-strip $(BIN)"
+
+container $(TEMP_DIR)/.container: $(TEMP_DIR)/$(BIN)
+	docker build -t $(REGISTRY)/pause-$(ARCH):$(TAG) $(TEMP_DIR)
 ifeq ($(ARCH),amd64)
 	docker tag -f $(REGISTRY)/pause-$(ARCH):$(TAG) $(REGISTRY)/pause:$(TAG)
 endif
+	touch $(TEMP_DIR)/.container
 
-push: build
+push $(TEMP_DIR)/.push: $(TEMP_DIR)/.container
 	gcloud docker --server=gcr.io push $(REGISTRY)/pause-$(ARCH):$(TAG)
```

Comment:
I don't think it's required to specify `--server=gcr.io` at all, since it isn't there in the most other Makefiles